File: src/runtime/boxes/dvar/bb_dvar.java
package bb;

/**
 * bb_dvar.java — DVAR: *VAR — re-resolve live variable value on every α
 * Port of bb_dvar.c / bb_dvar.s
 *
 * On each α entry, look up the variable's current value:
 *   - If DT_P (pattern-valued): rebuild child box from live pattern tree
 *   - If DT_S (string-valued):  rebuild child as bb_lit with that string
 * Then call child(α) / child(β) as normal.
 *
 *   DVAR_α:  resolve name → rebuild child if changed;
 *            DVAR = child(α);  if empty → DVAR_ω;  else → DVAR_γ;
 *   DVAR_β:  DVAR = child(β);  if empty → DVAR_ω;  else → DVAR_γ;
 */
public class bb_dvar extends bb_box {

    /** Callback to resolve variable → current bb_box (handles DT_P and DT_S) */
    public interface BoxResolver {
        /** Return a fresh bb_box for the current value of varname, or null if unset */
        bb_box resolve(String varname, MatchState ms);
    }

    private final String      varname;
    private final BoxResolver resolver;
    private       bb_box       child = null;

    public bb_dvar(MatchState ms, String varname, BoxResolver resolver) {
        super(ms);
        this.varname  = varname;
        this.resolver = resolver;
    }

    @Override public Spec α() {
        // DVAR_α: re-resolve on every fresh entry
        child = resolver.resolve(varname, ms);
        if (child == null) return null;                                        // DVAR_ω
        Spec r = child.α();
        return r;                                                              // DVAR_γ or DVAR_ω
    }

    @Override public Spec β() {
        // DVAR_β: delegate to same child (no re-resolve on backtrack)
        if (child == null) return null;                                        // DVAR_ω
        return child.β();
    }
}

File: src/runtime/boxes/dvar/bb_dvar.cs
// bb_dvar.cs — DVAR: *VAR — re-resolve variable's live value on every α
// Mirrors src/runtime/boxes/bb_dvar.c
//
// On α: read varname from IdentifierTable.
//   If value is a pattern (IByrdBox) → use it as child
//   If value is a string → wrap as bb_lit child
//   Re-build child only when value changes.
// α/β: delegate entirely to child.

namespace Snobol4.Runtime.Boxes;

public sealed class bb_dvar : IByrdBox
{
    private readonly string _varname;

    // Injected by executor
    public Func<string, string>?   GetStringVar   { get; set; }
    public Func<string, IByrdBox?>? GetPatternVar  { get; set; }

    private IByrdBox? _child;
    private string?   _lastStringValue;

    public bb_dvar(string varname) { _varname = varname ?? ""; }

    public Spec Alpha(MatchState ms)
    {
        RebuildChild();
        if (_child == null) return Spec.Fail;
        return _child.Alpha(ms);
    }

    public Spec Beta(MatchState ms)
    {
        if (_child == null) return Spec.Fail;
        return _child.Beta(ms);
    }

    private void RebuildChild()
    {
        // Try pattern value first
        var pat = GetPatternVar?.Invoke(_varname);
        if (pat != null)
        {
            if (!ReferenceEquals(pat, _child))
            {
                _child = pat;
                _lastStringValue = null;
            }
            return;
        }

        // Fall back to string value → wrap as literal
        var s = GetStringVar?.Invoke(_varname) ?? "";
        if (_child is bb_lit && s == _lastStringValue) return;  // unchanged
        _lastStringValue = s;
        _child = new bb_lit(s);
    }
}
